inference-engine/thirdparty/clDNN/src/network.cpp

```diff
@@ -393,23 +396,131 @@ void network_impl::set_input_data(const primitive_id& id, memory::ptr data) {
+    std::vector<std::shared_ptr<primitive_inst>> chain;
+    std::stack<std::shared_ptr<const primitive_inst>> candidates;
+    auto& eng = get_engine();
+    const auto& mem_orig = p_inst->output_memory();
 
-    if (primitive_inst == nullptr)
+    auto add_mdata_chain = [&](std::shared_ptr<primitive_inst>& p_inst) {
+        // special handling for mutable data, which can share
+        // its attached memory with both its inputs and outputs
+        for (auto& dep : p_inst->dependencies()) {
+            // check dependencies
+            if (eng.is_the_same_buffer(mem_orig, dep->output_memory())) {
+                chain.push_back(std::const_pointer_cast<primitive_inst>(dep));
+            }
+            // then second order dependencies
+            for (auto& second_dep : dep->dependencies()) {
+                if (eng.is_the_same_buffer(mem_orig, second_dep->output_memory())) {
+                    chain.push_back(std::const_pointer_cast<primitive_inst>(second_dep));
+                }
+            }
+        }
+
+        //then users
+        const auto& users = p_inst->get_users();
+        for (const auto& usr : users) {
+            auto usr_prim = get_primitive(usr->id());
+            if (eng.is_the_same_buffer(mem_orig, usr_prim->output_memory())) {
+                chain.push_back(usr_prim);
+            }
+        }
+    };
+
+    auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(p_inst);
+    if (p_inst->can_be_optimized()) {
+        candidates.push(p_inst);
+    } else if (mdata_ptr != nullptr) {
+        add_mdata_chain(p_inst);
+    }
+
+    // find all dependencies that are 'optimized'
+    while (!candidates.empty()) {
+        auto& cand = candidates.top();
+        candidates.pop();
+        const auto& mem_cand = cand->output_memory();
+        if (eng.is_the_same_buffer(mem_orig, mem_cand)) {
+            auto nc_cand = std::const_pointer_cast<primitive_inst>(cand);
+            chain.push_back(nc_cand);
+            auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(nc_cand);
+            if (mdata_ptr != nullptr) {
+                add_mdata_chain(nc_cand);
+            }
+        }
+
+        for (auto& dep : cand->dependencies()) {
+            if (dep->can_be_optimized()) {
+                candidates.push(dep);
+            } else {
+                const auto& mem_dep = dep->output_memory();
+                if (eng.is_the_same_buffer(mem_orig, mem_dep)) {
+                    auto nc_dep = std::const_pointer_cast<primitive_inst>(dep);
+                    chain.push_back(nc_dep);
+                    auto mdata_ptr = std::dynamic_pointer_cast<mutable_data_inst>(nc_dep);
+                    if (mdata_ptr != nullptr) {
+                        add_mdata_chain(nc_dep);
+                    }
+                }
+            }
+        }
+    }
+
+    if (chain.empty()) {
+        return _output_chains.end();
```

Comment:
I'd say that this method should never return end() iterator, as we always have at least one primitive that requires output update (p_inst)

Reply:
Fixed